Bug 1509214 - Convert Perfherder phCompareComponent

sarah-clements opened the pull request:
This code is for the table (actually a div of tables) for each test and associated platforms along with the filters and dropdown button for switching to a different framework that is used by the compare and comparesubtests views.

My apologies this is such a beast, but there were many intertwined parts that would've made it difficult to do in smaller chunks.

I'll experiment with shallow enzyme tests in another branch... I meant to start on that sooner :)

@wlach I'd appreciate your feedback but no worries if you don't have time.

wlach commented on ui/perfherder/TableAverage.jsx:
```diff
@@ -0,0 +1,68 @@
+  const notZeroSum =
+    replicates.length > 1
+      ? replicates.reduce((accumulator, current) => accumulator + current)
+      : false;
```
Could simplify this slightly to:

`const notZeroSum = replicates.length && replicates.reduce((accumulator, current) => accumulator + current)`

wlach commented on ui/perfherder/TableAverage.jsx:
```diff
@@ -0,0 +1,68 @@
+              tooltipText
+            )
+          }
+          tooltipClass={replicates.length > 1 ? 'compare-table-tooltip' : ''}
```
`replicates.length` does the same job as `replicates.length > 1` here.

wlach commented on ui/intermittent-failures/BugDetailsView.jsx:
```diff
@@ -76,7 +80,44 @@ const BugDetailsView = props => {
+            }
+            placement="left"
+            tooltipText={
+              original.lines.length > 0 && (
```
As above

wlach commented on ui/intermittent-failures/BugDetailsView.jsx:
```diff
@@ -76,7 +80,44 @@ const BugDetailsView = props => {
+            text={
+              <React.Fragment>
+                {`${original.lines.length} unexpected-fail${
+                  original.lines.length > 1 ? 's' : ''
```
`original.lines.length > 0` => `original.lines.length`

sarah-clements replied:
I'm not understanding this suggestion. The existing syntax is correct - return an 's' if there's more than one so the text for "1 unexpected fail" is not pluralized.

wlach commented:
Did some testing, everything looks good to me. There are some things which are less than perfect about this view, which would be nice to clean up at some point (i.e. the ui doesn't tell you whether something is worse or better except via a hard-to-discover tooltip and some of the verbiage is a bit misleading) but I think it's best to start with a 1:1 port and then figure out what can be improved later, as you just did.

wlach commented:
Actually, one thing I'm not familiar with but may also be worth checking is the noise metric stuff, which I've mostly forgotten about. May want to check with @jmaher or @klahnakoski about whether that is still working as expected

sarah-clements commented:
I forgot to mention that I made some design and color scheme changes: replaced form element check boxes with react-straps toggle buttons, and replaced colored backgrounds with icons (to give a visual cue for red-green colorblind people). Not sure about keeping "the warning yellow" as a background color. Any opinions on this and the choice of icons? Could do thumbs up, thumbs down and ! for yellow warning instead? @wlach @ionutgoldan 

![screen shot 2019-02-01 at 12 31 02 pm](https://user-images.githubusercontent.com/19615783/52148071-6a400b80-261d-11e9-9316-c114a0c07804.png)

sarah-clements commented:
> Did some testing, everything looks good to me. There are some things which are less than perfect about this view, which would be nice to clean up at some point (i.e. the ui doesn't tell you whether something is worse or better except via a hard-to-discover tooltip and some of the verbiage is a bit misleading) but I think it's best to start with a 1:1 port and then figure out what can be improved later, as you just did.

yeah, I'll be coordinating with the Perf team in future on improvements I can make going forward (realizing after the fact there were some changes that are desired).

wlach commented:
> replaced form element check boxes with react-straps toggle buttons

Forgot to mention that. This seems like the one change that I'm less certain about. I feel like checkboxes are a little clearer / less ambiguous than toggle buttons, even if slightly less pretty. 

> replaced colored backgrounds with icons (to give a visual cue for red-green colorblind people)

This is better for people who are using screen readers too. 👍 

> Not sure about keeping "the warning yellow" as a background color. Any opinions on this and the choice of icons? Could do thumbs up, thumbs down and ! for yellow warning instead?

I don't really have a strong opinion. I think I chose yellow for "medium confidence" results to indicate that "people may want to have a look at this but it's probably not a big deal". I think it's pretty rare for these kinds of results on their own indicate much of anything on their own-- any time you're comparing a lot of different things, you'll notice a moderately significant difference in a few cases. I would lean towards leaving this as-is for now, maybe revisiting later.

sarah-clements commented:
> > replaced form element check boxes with react-straps toggle buttons
> 
> Forgot to mention that. This seems like the one change that I'm less certain about. I feel like checkboxes are a little clearer / less ambiguous than toggle buttons, even if slightly less pretty.

I see what you mean. Perhaps going with a traditional looking button would make it less ambiguous?

![screen shot 2019-02-01 at 1 16 35 pm](https://user-images.githubusercontent.com/19615783/52150139-e2a9cb00-2623-11e9-93d3-fb811349128e.png)

sarah-clements commented:
@wlach Thanks for the feedback!

sarah-clements commented on ui/js/controllers/perf/compare.js:
```diff
@@ -127,29 +126,35 @@ perf.controller('CompareResultsCtrl', [
                     cmap.name = platform;
-                    if (Object.keys($scope.compareResults).indexOf(testName) < 0) {
-                        $scope.compareResults[testName] = [];
+                    if ($scope.compareResults.has(testName)) {
```
yeah, I was going to tackle that when I work on the compare view and the comparesubtests views since there is a lot of logic that can be reused.

sarah-clements commented:
> Looks great! The only thing I noticed in testing was when I typed in a filter and hit enter, I expected it to do the filtering, rather than having to click the filter button. Minor thing that could be addressed in another PR, though.

Its not ideal, I know. I was experiencing lag when typing in the input due to the constant state change causing everything to re-render in quick succession. So I added the button and moved the input state into its own component. Happy to try a different approach if you know of one :)

camd commented:
> > Looks great! The only thing I noticed in testing was when I typed in a filter and hit enter, I expected it to do the filtering, rather than having to click the filter button. Minor thing that could be addressed in another PR, though.
> 
> Its not ideal, I know. I was experiencing lag/performance issues when typing in the input due to the constant state change causing everything to re-render in quick succession. So I added the button and moved the input state into its own component. Happy to try a different approach if you know of one :)



> > Looks great! The only thing I noticed in testing was when I typed in a filter and hit enter, I expected it to do the filtering, rather than having to click the filter button. Minor thing that could be addressed in another PR, though.
> 
> Its not ideal, I know. I was experiencing lag/performance issues when typing in the input due to the constant state change causing everything to re-render in quick succession. So I added the button and moved the input state into its own component. Happy to try a different approach if you know of one :)

The only other approach I could think of would be to make that field and button a ``Form`` and have the button be a ``submit`` button.  tbh, I'm not entirely sure I've done this with a form that wasn't a django submit form, so maybe the React one won't work that way.  Anyway, it could definitely be added at a later time.  Shouldn't hold this PR up.  :)

sarah-clements commented:
> The only other approach I could think of would be to make that field and button a `Form` and have the button be a `submit` button. tbh, I'm not entirely sure I've done this with a form that wasn't a django submit form, so maybe the React one won't work that way. Anyway, it could definitely be added at a later time. Shouldn't hold this PR up. :)

Ah, I see what you mean - thanks for pointing this out. I misread your comment and thought you were expecting on-the-fly filtering. I think using `onKeyPress` on the input element would work.

(sorry, accidentally edited your original comment!)

wlach commented:
> > Looks great! The only thing I noticed in testing was when I typed in a filter and hit enter, I expected it to do the filtering, rather than having to click the filter button. Minor thing that could be addressed in another PR, though.
> 
> Its not ideal, I know. I was experiencing lag/performance issues when typing in the input due to the constant state change causing everything to re-render in quick succession. So I added the button and moved the input state into its own component. Happy to try a different approach if you know of one :)

I can't remember the incantation to do this with react, but you should be able to debounce the input so changes only take effect after a short delay. Googling for "debounce react" turns up a bunch of results.

> I'm not opposed to checkboxes in general :) This article sums up the apprehension with using checkboxes with the current behavior. The tldr; Tapping a toggle switch is a two-step action: selection and execution, whereas checkbox is just selection of an option and its execution usually requires another control.

Hmm, I'm not sure if I agree with that article -- I personally think it's perfectly fine and intuitive for checkboxes to be instant-apply. That said, I think this is your decision to make and won't argue it further.

sarah-clements commented:
@jmaher Can you let me know if the noise metric is still behaving as expected? (you can try it out on [prototype](https://treeherder-prototype.herokuapp.com/perf.html#/compare?originalProject=mozilla-central&newProject=try&newRevision=2a2b3f6fa82f72bdbdf9fab0a268095d1d966d8a&framework=1&selectedTimeRange=172800))

edmorley commented on ui/perfherder/CompareTable.jsx:
```diff
@@ -0,0 +1,198 @@
+      <Table sz="small" className="compare-table mb-0" key={title || testName}>
+        <thead>
+          <tr
+            style={{ backgroundColor: 'lightgrey' }}
```
Inline styles don't work with CSP (bug 1270157) - could you convert these to CSS classes whilst you're here? (Some will be supported by bootstrap utilities so may not need new class definitions)

sarah-clements replied:
I've been using inline styles in all new Perf components, so I'll open another bug to change those to classes.

edmorley commented on ui/perfherder/CompareTable.jsx:
```diff
@@ -0,0 +1,198 @@
+                    text={
+                      <React.Fragment>
+                        {(results.isRegression || results.isImprovement) && (
+                          <Icon
```
Using the new `@fortawesome/react-fontawesome` package, this would be something like:

```js
<FontAwesomeIcon
  icon={results.isRegression ? faExclamationTriangle : faThumbsUp}
  size="..."  // replaces `fontSize`, see https://fontawesome.com/how-to-use/on-the-web/styling/sizing-icons
  // other attributes stay the same
/>
```

sarah-clements commented on ui/perfherder/CompareTable.jsx:
```diff
@@ -0,0 +1,198 @@
+import React from 'react';
+import { Table, Progress } from 'reactstrap';
+import PropTypes from 'prop-types';
+import Icon from 'react-fontawesome';
```
Done - thanks!

sarah-clements commented:
@edmorley I've made the requested changes - please approve.
